subsys/nfc/tnep/poller.c

```diff
@@ -0,0 +1,611 @@
+}
+
+int nfc_tnep_poller_init(const struct nfc_tnep_buff *tx_buff,
+			 const struct nfc_tnep_poller_cb *cb)
+{
+	if (!tx_buff || !cb) {
+		return -EINVAL;
+	}
+
+	if (!tx_buff->data) {
+		LOG_ERR("NULL buffer provided");
+		return -EINVAL;
+	}
+
+	if (tx_buff->size < 1) {
+		LOG_ERR("Tx Buffer length cannot have length equal to 0");
+		return -ENOMEM;
+	}
+
+	tnep.cb = cb;
+	tnep.tx = tx_buff;
+	tnep.state = TNEP_POLLER_STATE_IDLE;
+
+	k_delayed_work_init(&tnep.tnep_work, tnep_delay_handler);
+
+	return 0;
+}
+
+int nfc_tnep_poller_api_set(const struct nfc_tnep_poller_ndef_api *api,
+			    enum nfc_tnep_tag_type tag_type)
+{
+	if (!api) {
+		return -EINVAL;
+	}
+
+	if ((tag_type != NFC_TNEP_TAG_TYPE_T2T) &&
+	    (tag_type != NFC_TNEP_TAG_TYPE_T4T)) {
+		LOG_ERR("Unknown NFC Tag Type");
+		return -EINVAL;
+	}
+
+	tnep.api = api;
+	tnep.type = tag_type;
+
+	return 0;
+}
+
+int nfc_tnep_poller_svc_search(const struct nfc_ndef_msg_desc *ndef_msg,
+			       struct nfc_ndef_tnep_rec_svc_param *params,
+			       u8_t *cnt)
+{
+	int err;
+	const struct nfc_ndef_record_desc *record;
+	u8_t size;
+
+	if (!ndef_msg || !params || !cnt) {
+		LOG_ERR("NULL argument.");
+		return -EINVAL;
+	}
+
+	if ((*cnt) < 1) {
+		return -ENOMEM;
+	}
+
+	size = *cnt;
+	*cnt = 0;
+
+	if (ndef_msg->record_count < 1) {
+		LOG_DBG("No NDEF Records in message.");
+		return 0;
+	}
+
+	/* Look for TNEP Service Parameters Records. */
+	for (size_t i = 0; i < ndef_msg->record_count; i++) {
+		record = ndef_msg->record[i];
+
+		/* Check if record is Service Parameters Record. */
+		if (service_param_rec_check(record->tnf, record->type,
+					    record->type_length)) {
+			/* Check if we have memory for service. */
+			if ((*cnt) > size) {
+				return -ENOMEM;
+			}
+
+			/* If service parameters record corrupted.
+			 * Search for next valid record.
+			 */
+			err = svc_param_record_get(record->payload_descriptor,
+						   &params[*cnt]);
+			if (err) {
+				LOG_DBG("Service Parameter Record corrupted.");
+				continue;
+			}
+
+			(*cnt)++;
+		}
+	}
+
+	return 0;
+}
+
+int nfc_tnep_poller_svc_select(u8_t *svc_buff, size_t buff_len,
+			       const struct nfc_ndef_tnep_rec_svc_param *svc,
+			       u32_t max_ndef_area_size)
+{
+	int err;
+	size_t len = 0;
+
+	NFC_NDEF_MSG_DEF(select_msg, TNEP_SELECT_MSG_RECORD_CNT);
+
+	if (!svc || !svc_buff) {
+		return -EINVAL;
+	}
+
+	if (buff_len < 1) {
+		return -ENOMEM;
+	}
+
+	tnep.state = TNEP_POLLER_STATE_SELECTING;
+
+	/* If service is currently selected do not select it twice. */
+	if (svc == tnep.active_svc) {
+		LOG_DBG("Service currently selected. Reading data");
+
+		on_svc_delayed_operation();
+
+		return 0;
+	}
+
+	if (svc->communication_mode != NFC_TNEP_COMM_MODE_SINGLE_RESPONSE) {
+		LOG_DBG("Unsupported communication mode");
+		return -ENOTSUP;
+	}
+
+	if (svc->max_size > max_ndef_area_size) {
+		LOG_DBG("Service NDEF max size greater than max NDEF size. Service message size should be smaller than %d for this service",
```

Comment:
please break this line at the start of the second sentence

Reply:
Ok